File: BubbleLab/services/openevolve-api/models/team_assignment.py
"""
Flexible Team LLM Assignment System

Allows arbitrary LLM/vLLM assignment to any team (blue/red/judge) with
unified credential management from OpenEvolve config and BubbleLab credentials.

Key Features:
- Arbitrary LLM assignment to teams
- vLLM distinction for visual workflows
- Unified credential management
- Credential verification
- Team composition flexibility
"""

from pydantic import BaseModel, Field, validator
from typing import Dict, List, Optional, Union, Literal
from enum import Enum


class LLMProvider(str, Enum):
    """Supported LLM providers"""
    OPENAI = "openai"
    ANTHROPIC = "anthropic"
    GOOGLE = "google"
    OPENROUTER = "openrouter"
    TOGETHER = "together"
    GROQ = "groq"
    DEEPSEEK = "deepseek"
    OPENAI_LIKE = "openai-like"  # Compatible APIs (vLLM, Ollama, etc.)
    CUSTOM = "custom"


class LLMCapability(str, Enum):
    """LLM capabilities"""
    TEXT = "text"  # Text generation only
    VISION = "vision"  # Multimodal (vLLM)
    CODE = "code"  # Code generation optimized
    MATH = "math"  # Mathematical reasoning
    REASONING = "reasoning"  # Complex reasoning
    TOOL_USE = "tool_use"  # Function calling
    AGENTIC = "agentic"  # Agentic behavior


class LLMModel(BaseModel):
    """
    LLM model definition with capabilities
    """
    provider: LLMProvider
    model_id: str = Field(..., description="Model identifier (e.g., 'gpt-4', 'claude-3-opus')")
    name: str = Field(..., description="Human-readable model name")
    capabilities: List[LLMCapability] = Field(
        default_factory=lambda: [LLMCapability.TEXT],
        description="Model capabilities"
    )
    max_tokens: int = Field(default=4096, ge=1, le=1000000)
    supports_streaming: bool = Field(default=True)
    supports_function_calling: bool = Field(default=False)
    is_vision: bool = Field(default=False, description="Is this a vision/multimodal model?")

    # Optional pricing
    input_price_per_1k: Optional[float] = Field(None, description="Input price per 1K tokens")
    output_price_per_1k: Optional[float] = Field(None, description="Output price per 1K tokens")

    @validator('capabilities')
    def validate_capability_consistency(cls, v, values):
        """Ensure vision capability is set if is_vision is True"""
        if values.get('is_vision') and LLMCapability.VISION not in v:
            v.append(LLMCapability.VISION)
        return v


class CredentialSource(str, Enum):
    """Where credentials come from"""
    OPENEVOLVE_CONFIG = "openevolve_config"  # From OpenEvolve .env/config
    BUBBLELAB_CREDENTIALS = "bubblelab_credentials"  # From BubbleLab credentials API
    USER_PROVIDED = "user_provided"  # Provided at runtime


class LLMCredential(BaseModel):
    """
    LLM credentials with verification status
    """
    credential_id: str
    provider: LLMProvider
    api_key: str = Field(..., description="Encrypted API key")
    source: CredentialSource
    verified: bool = Field(default=False)
    verified_at: Optional[str] = None  # ISO 8601 timestamp
    last_used: Optional[str] = None  # ISO 8601 timestamp
    model_permissions: List[str] = Field(
        default_factory=list,
        description="Models this credential can access"
    )

    # Optional additional auth parameters
    api_base: Optional[str] = Field(None, description="Custom API base URL (for OpenAI-compatible APIs)")
    organization_id: Optional[str] = Field(None, description="Organization ID (for OpenAI)")
    project_id: Optional[str] = Field(None, description="Project ID (for Google/Vertex)")
    region: Optional[str] = Field(None, description="Region (for some providers)")


class TeamRole(str, Enum):
    """Standard team roles"""
    BLUE_TEAM = "blue"  # Generates solutions
    RED_TEAM = "red"  # Attacks/evaluates solutions
    JUDGE = "judge"  # Evaluates and decides
    OBSERVER = "observer"  # Watches and learns
    ARBITER = "arbitrator"  # Resolves disagreements


class TeamMemberLLM(BaseModel):
    """
    LLM assigned to a team
    """
    member_id: str
    llm: LLMModel
    credential_id: Optional[str] = Field(None, description="Credential to use (None = use default)")
    role: TeamRole = Field(..., description="Team role")
    temperature: float = Field(default=0.7, ge=0.0, le=2.0)
    max_tokens: int = Field(default=4096, ge=1, le=1000000)

    # Optional personality/prompt customization
    system_prompt: Optional[str] = Field(
        None,
        description="Custom system prompt for this LLM in this team context"
    )
    personality: Optional[str] = Field(
        None,
        description="Personality preset (e.g., 'critical', 'creative', 'conservative')"
    )

    # Performance tracking
    total_requests: int = Field(default=0)
    successful_requests: int = Field(default=0)
    average_latency_ms: Optional[float] = Field(None)


class Team(BaseModel):
    """
    A team with multiple LLM members
    """
    team_id: str
    name: str
    description: Optional[str] = None
    members: List[TeamMemberLLM] = Field(
        default_factory=list,
        description="LLMs assigned to this team"
    )

    # Team strategy
    voting_strategy: Literal["consensus", "majority", "weighted", "leader_decides"] = "consensus"
    quorum_threshold: float = Field(
        default=0.7,
        ge=0.0,
        le=1.0,
        description="Fraction of team needed for decision"
    )

    # Team composition rules
    require_vision_for_design: bool = Field(
        default=True,
        description="Require at least one vLLM for design tasks"
    )
    require_diverse_providers: bool = Field(
        default=False,
        description="Require LLMs from different providers"
    )


class TeamAssignmentRequest(BaseModel):
    """
    Request to assign LLM to a team
    """
    team_id: str
    llm_provider: LLMProvider
    llm_model_id: str
    role: TeamRole
    temperature: float = 0.7
    max_tokens: int = 4096
    credential_id: Optional[str] = None
    system_prompt: Optional[str] = None


class CredentialVerificationRequest(BaseModel):
    """
    Request to verify LLM credentials
    """
    credential_id: Optional[str] = Field(None, description="Existing credential to verify")
    provider: LLMProvider
    api_key: str = Field(..., description="API key to verify")
    api_base: Optional[str] = Field(None, description="Custom API base URL")
    model_to_test: Optional[str] = Field(None, description="Model to test credential with")


class CredentialVerificationResponse(BaseModel):
    """
    Result of credential verification
    """
    verified: bool
    credential_id: Optional[str] = None
    message: str
    test_model: Optional[str] = None
    latency_ms: Optional[float] = None
    available_models: Optional[List[str]] = None


# Predefined LLM catalog with capabilities
PREDEFINED_LLM_MODELS = {
    # OpenAI
    "gpt-4": LLMModel(
        provider=LLMProvider.OPENAI,
        model_id="gpt-4",
        name="GPT-4",
        capabilities=[LLMCapability.TEXT, LLMCapability.CODE, LLMCapability.REASONING],
        supports_function_calling=True,
        input_price_per_1k=0.03,
        output_price_per_1k=0.06,
    ),
    "gpt-4-turbo": LLMModel(
        provider=LLMProvider.OPENAI,
        model_id="gpt-4-turbo-preview",
        name="GPT-4 Turbo",
        capabilities=[LLMCapability.TEXT, LLMCapability.CODE, LLMCapability.REASONING, LLMCapability.VISION],
        is_vision=True,
        supports_function_calling=True,
        input_price_per_1k=0.01,
        output_price_per_1k=0.03,
    ),
    "gpt-4-vision": LLMModel(
        provider=LLMProvider.OPENAI,
        model_id="gpt-4-vision-preview",
        name="GPT-4 Vision",
        capabilities=[LLMCapability.TEXT, LLMCapability.VISION],
        is_vision=True,
        input_price_per_1k=0.01,
        output_price_per_1k=0.03,
    ),
    "gpt-3.5-turbo": LLMModel(
        provider=LLMProvider.OPENAI,
        model_id="gpt-3.5-turbo",
        name="GPT-3.5 Turbo",
        capabilities=[LLMCapability.TEXT, LLMCapability.CODE],
        supports_function_calling=True,
        input_price_per_1k=0.0005,
        output_price_per_1k=0.0015,
    ),

    # Anthropic
    "claude-3-opus": LLMModel(
        provider=LLMProvider.ANTHROPIC,
        model_id="claude-3-opus-20240229",
        name="Claude 3 Opus",
        capabilities=[LLMCapability.TEXT, LLMCapability.CODE, LLMCapability.REASONING, LLMCapability.VISION],
        is_vision=True,
        max_tokens=200000,
        supports_function_calling=True,  # Claude 3 supports tool use
        input_price_per_1k=0.015,
        output_price_per_1k=0.075,
    ),
    "claude-3-sonnet": LLMModel(
        provider=LLMProvider.ANTHROPIC,
        model_id="claude-3-sonnet-20240229",
        name="Claude 3 Sonnet",
        capabilities=[LLMCapability.TEXT, LLMCapability.CODE, LLMCapability.VISION],
        is_vision=True,
        max_tokens=200000,
        input_price_per_1k=0.003,
        output_price_per_1k=0.015,
    ),
    "claude-3-haiku": LLMModel(
        provider=LLMProvider.ANTHROPIC,
        model_id="claude-3-haiku-20240307",
        name="Claude 3 Haiku",
        capabilities=[LLMCapability.TEXT, LLMCapability.CODE, LLMCapability.VISION],
        is_vision=True,
        max_tokens=200000,
        input_price_per_1k=0.00025,
        output_price_per_1k=0.00125,
    ),

    # Google
    "gemini-pro": LLMModel(
        provider=LLMProvider.GOOGLE,
        model_id="gemini-pro",
        name="Gemini Pro",
        capabilities=[LLMCapability.TEXT, LLMCapability.REASONING, LLMCapability.VISION],
        is_vision=True,
        input_price_per_1k=0.00025,
        output_price_per_1k=0.0005,
    ),
    "gemini-ultra": LLMModel(
        provider=LLMProvider.GOOGLE,
        model_id="gemini-ultra",
        name="Gemini Ultra",
        capabilities=[LLMCapability.TEXT, LLMCapability.CODE, LLMCapability.REASONING, LLMCapability.VISION],
        is_vision=True,
        input_price_per_1k=0.00175,
        output_price_per_1k=0.0021,
    ),

    # Groq (Fast inference)
    "llama3-70b-groq": LLMModel(
        provider=LLMProvider.GROQ,
        model_id="llama3-70b-8192",
        name="Llama 3 70B (Groq)",
        capabilities=[LLMCapability.TEXT, LLMCapability.CODE, LLMCapability.REASONING],
        supports_streaming=True,
        input_price_per_1k=0.0000?  # Groq pricing varies
        output_price_per_1k=0.0000?,
    ),

    # DeepSeek
    "deepseek-coder": LLMModel(
        provider=LLMProvider.DEEPSEEK,
        model_id="deepseek-coder",
        name="DeepSeek Coder",
        capabilities=[LLMCapability.TEXT, LLMCapability.CODE],
        max_tokens=4096,
    ),

    # OpenRouter (aggregator)
    "openrouter-mix": LLMModel(
        provider=LLMProvider.OPENROUTER,
        model_id="anthropic/claude-3-opus",
        name="Claude 3 Opus (via OpenRouter)",
        capabilities=[LLMCapability.TEXT, LLMCapability.CODE, LLMCapability.REASONING],
    ),
}


def get_available_llms(
    provider: Optional[LLMProvider] = None,
    capability: Optional[LLMCapability] = None,
    vision_only: bool = False,
) -> List[LLMModel]:
    """
    Get available LLMs with optional filtering

    Args:
        provider: Filter by provider
        capability: Filter by capability
        vision_only: Only return vLLMs

    Returns:
        List of available LLM models
    """
    llms = list(PREDEFINED_LLM_MODELS.values())

    if provider:
        llms = [llm for llm in llms if llm.provider == provider]

    if capability:
        llms = [llm for llm in llms if capability in llm.capabilities]

    if vision_only:
        llms = [llm for llm in llms if llm.is_vision]

    return llms


def get_vision_llms() -> List[LLMModel]:
    """Get all vLLMs (models with vision capability)"""
    return get_available_llms(vision_only=True)


def group_llms_by_capability() -> Dict[str, List[LLMModel]]:
    """
    Group LLMs by capability for UI organization

    Returns:
        Dict mapping capability names to lists of LLMs
    """
    groups = {}

    for llm in PREDEFINED_LLM_MODELS.values():
        if llm.is_vision:
            key = "Vision/Multimodal (vLLM)"
        elif LLMCapability.CODE in llm.capabilities:
            key = "Code Generation"
        elif LLMCapability.REASONING in llm.capabilities:
            key = "Reasoning & Analysis"
        else:
            key = "General Purpose"

        if key not in groups:
            groups[key] = []
        groups[key].append(llm)

    return groups

File: BubbleLab/apps/bubble-studio/src/types/team-assignment.ts
/**
 * Flexible Team LLM Assignment Types
 *
 * Supports arbitrary LLM/vLLM assignment to any team
 * with unified credential management
 */

// ==================== LLM Provider Types ====================

export enum LLMProvider {
  OPENAI = 'openai',
  ANTHROPIC = 'anthropic',
  GOOGLE = 'google',
  OPENROUTER = 'openrouter',
  TOGETHER = 'together',
  GROQ = 'groq',
  DEEPSEEK = 'deepseek',
  OPENAI_LIKE = 'openai-like', // vLLM, Ollama, etc.
  CUSTOM = 'custom',
}

export enum LLMCapability {
  TEXT = 'text',
  VISION = 'vision', // vLLM capability
  CODE = 'code',
  MATH = 'math',
  REASONING = 'reasoning',
  TOOL_USE = 'tool_use',
  AGENTIC = 'agentic',
}

export enum CredentialSource {
  OPENEVOLVE_CONFIG = 'openevolve_config', // From .env file
  BUBBLELAB_CREDENTIALS = 'bubblelab_credentials', // From BubbleLab API
  USER_PROVIDED = 'user_provided', // Entered by user at runtime
}

export enum TeamRole {
  BLUE_TEAM = 'blue', // Generates solutions
  RED_TEAM = 'red', // Attacks/evaluates
  JUDGE = 'judge', // Evaluates and decides
  OBSERVER = 'observer', // Watches and learns
  ARBITER = 'arbitrator', // Resolves disputes
}

// ==================== Core Types ====================

export interface LLMModel {
  provider: LLMProvider;
  model_id: string;
  name: string;
  capabilities: LLMCapability[];
  max_tokens: number;
  supports_streaming: boolean;
  supports_function_calling: boolean;
  is_vision: boolean; // True for vLLMs
  input_price_per_1k?: number;
  output_price_per_1k?: number;
}

export interface LLMCredential {
  credential_id: string;
  provider: LLMProvider;
  api_key: string; // Will be encrypted in storage
  source: CredentialSource;
  verified: boolean;
  verified_at?: string; // ISO 8601
  last_used?: string; // ISO 8601
  model_permissions: string[];
  api_base?: string; // For OpenAI-compatible APIs
  organization_id?: string;
  project_id?: string;
  region?: string;
}

export interface TeamMemberLLM {
  member_id: string;
  llm: LLMModel;
  credential_id?: string;
  role: TeamRole;
  temperature: number;
  max_tokens: number;
  system_prompt?: string;
  personality?: string;
  total_requests: number;
  successful_requests: number;
  average_latency_ms?: number;
}

export interface Team {
  team_id: string;
  name: string;
  description?: string;
  members: TeamMemberLLM[];
  voting_strategy: 'consensus' | 'majority' | 'weighted' | 'leader_decides';
  quorum_threshold: number; // 0.0 to 1.0
  require_vision_for_design: boolean;
  require_diverse_providers: boolean;
}

export interface TeamAssignmentRequest {
  team_id: string;
  llm_provider: LLMProvider;
  llm_model_id: string;
  role: TeamRole;
  temperature: number;
  max_tokens: number;
  credential_id?: string;
  system_prompt?: string;
}

export interface CredentialVerificationRequest {
  credential_id?: string;
  provider: LLMProvider;
  api_key: string;
  api_base?: string;
  model_to_test?: string;
}

export interface CredentialVerificationResponse {
  verified: boolean;
  credential_id?: string;
  message: string;
  test_model?: string;
  latency_ms?: number;
  available_models?: string[];
}

// ==================== UI Types ====================

export interface LLMSearchFilters {
  provider?: LLMProvider;
  capability?: LLMCapability;
  vision_only?: boolean;
  search_query?: string;
}

export interface LLMGroup {
  category: string;
  llms: LLMModel[];
}

export interface TeamComposition {
  team_id: string;
  name: string;
  members: TeamMemberLLM[];
  has_vision: boolean;
  provider_diversity: number; // Number of unique providers
  total_cost_estimate?: number; // Estimated cost per 1K tokens
}

export interface CredentialFormData {
  provider: LLMProvider;
  api_key: string;
  api_base?: string;
  organization_id?: string;
  project_id?: string;
  region?: string;
}

// ==================== Response Types ====================

export interface LLMSearchResponse {
  llms: LLMModel[];
  grouped: Record<string, LLMModel[]>;
  total: number;
  vision_llms: LLMModel[];
  text_llms: LLMModel[];
}

export interface CredentialsListResponse {
  credentials: Array<{
    credential_id: string;
    provider: LLMProvider;
    source: CredentialSource;
    verified: boolean;
    // API key is masked
    api_key_preview: string;
  }>;
  total: number;
  sources: CredentialSource[];
}

export interface TeamCreateResponse {
  team_id: string;
  name: string;
  members: TeamMemberLLM[];
  created_at: string;
}

export interface TeamTemplate {
  id: string;
  name: string;
  description: string;
  composition: Array<{
    role: TeamRole;
    llm: string; // model_id
    count: number;
  }>;
}

// ==================== Helper Types ====================

export type LLMProviderOption = {
  value: LLMProvider;
  label: string;
  vision_support: boolean;
};

export type TeamRoleOption = {
  value: TeamRole;
  label: string;
  description: string;
  color: string; // For UI
};

export type CapabilityBadge = {
  capability: LLMCapability;
  label: string;
  color: string; // For UI
};

// ==================== Default Values ====================

export const DEFAULT_TEAM_COMPOSITION = {
  voting_strategy: 'consensus' as const,
  quorum_threshold: 0.7,
  require_vision_for_design: true,
  require_diverse_providers: false,
};

export const DEFAULT_MEMBER_CONFIG = {
  temperature: 0.7,
  max_tokens: 4096,
};

// ==================== UI Constants ====================

export const TEAM_ROLE_OPTIONS: TeamRoleOption[] = [
  {
    value: TeamRole.BLUE_TEAM,
    label: 'Blue Team',
    description: 'Generates solutions',
    color: '#3b82f6', // blue
  },
  {
    value: TeamRole.RED_TEAM,
    label: 'Red Team',
    description: 'Attacks and evaluates',
    color: '#ef4444', // red
  },
  {
    value: TeamRole.JUDGE,
    label: 'Judge',
    description: 'Evaluates and decides',
    color: '#8b5cf6', // purple
  },
  {
    value: TeamRole.OBSERVER,
    label: 'Observer',
    description: 'Watches and learns',
    color: '#6b7280', // gray
  },
  {
    value: TeamRole.ARBITER,
    label: 'Arbiter',
    description: 'Resolves disagreements',
    color: '#f59e0b', // amber
  },
];

export const CAPABILITY_BADGES: Record<LLMCapability, CapabilityBadge> = {
  [LLMCapability.TEXT]: {
    capability: LLMCapability.TEXT,
    label: 'Text',
    color: '#6b7280',
  },
  [LLMCapability.VISION]: {
    capability: LLMCapability.VISION,
    label: 'Vision',
    color: '#10b981', // green for vLLM
  },
  [LLMCapability.CODE]: {
    capability: LLMCapability.CODE,
    label: 'Code',
    color: '#3b82f6',
  },
  [LLMCapability.MATH]: {
    capability: LLMCapability.MATH,
    label: 'Math',
    color: '#8b5cf6',
  },
  [LLMCapability.REASONING]: {
    capability: LLMCapability.REASONING,
    label: 'Reasoning',
    color: '#f59e0b',
  },
  [LLMCapability.TOOL_USE]: {
    capability: LLMCapability.TOOL_USE,
    label: 'Tools',
    color: '#ec4899',
  },
  [LLMCapability.AGENTIC]: {
    capability: LLMCapability.AGENTIC,
    label: 'Agentic',
    color: '#14b8a6',
  },
};

// ==================== Helper Functions ====================

export function isVisionLLM(llm: LLMModel): boolean {
  return llm.is_vision || llm.capabilities.includes(LLMCapability.VISION);
}

export function getLLMProviderLabel(provider: LLMProvider): string {
  const labels: Record<LLMProvider, string> = {
    [LLMProvider.OPENAI]: 'OpenAI',
    [LLMProvider.ANTHROPIC]: 'Anthropic',
    [LLMProvider.GOOGLE]: 'Google',
    [LLMProvider.OPENROUTER]: 'OpenRouter',
    [LLMProvider.TOGETHER]: 'Together',
    [LLMProvider.GROQ]: 'Groq',
    [LLMProvider.DEEPSEEK]: 'DeepSeek',
    [LLMProvider.OPENAI_LIKE]: 'OpenAI-Compatible',
    [LLMProvider.CUSTOM]: 'Custom',
  };
  return labels[provider] || provider;
}

export function formatCredentialSource(source: CredentialSource): string {
  const labels: Record<CredentialSource, string> = {
    [CredentialSource.OPENEVOLVE_CONFIG]: 'OpenEvolve Config',
    [CredentialSource.BUBBLELAB_CREDENTIALS]: 'Saved Credentials',
    [CredentialSource.USER_PROVIDED]: 'User Provided',
  };
  return labels[source] || source;
}

export function estimateCost(
  llm: LLMModel,
  inputTokens: number,
  outputTokens: number,
): number {
  const inputCost = (inputTokens / 1000) * (llm.input_price_per_1k || 0);
  const outputCost = (outputTokens / 1000) * (llm.output_price_per_1k || 0);
  return inputCost + outputCost;
}

export function groupLLMsByCapability(llms: LLMModel[]): Record<string, LLMModel[]> {
  const groups: Record<string, LLMModel[]> = {};

  for (const llm of llms) {
    let key: string;

    if (llm.is_vision) {
      key = 'Vision/Multimodal (vLLM)';
    } else if (llm.capabilities.includes(LLMCapability.CODE)) {
      key = 'Code Generation';
    } else if (llm.capabilities.includes(LLMCapability.REASONING)) {
      key = 'Reasoning & Analysis';
    } else {
      key = 'General Purpose';
    }

    if (!groups[key]) {
      groups[key] = [];
    }
    groups[key].push(llm);
  }

  return groups;
}
